// ngx-renderer/dist/renderer/noop/advanced-renderer.js
import { Injectable } from '@angular/core';
import { AdvancedRenderer } from '../advanced-renderer';
export class AdvancedRendererNoop extends AdvancedRenderer {
    /**
     * @param {?} expression
     * @param {?} args
     * @return {?}
     */
    execute(expression, args) {
        return Promise.resolve();
    }
    /**
     * @param {?} element
     * @param {?} method
     * @param {?} args
     * @return {?}
     */
    invokeElementMethod(element, method, args) {
        return Promise.resolve();
    }
    /**
     * @param {?} target
     * @param {?} method
     * @param {?} args
     * @return {?}
     */
    invokeGlobalTargetMethod(target, method, args) {
        return Promise.resolve();
    }
    /**
     * @param {?} target
     * @param {?} prop
     * @param {?} value
     * @return {?}
     */
    setGlobalTargetPropery(target, prop, value) { }
    /**
     * @param {?} target
     * @param {?} prop
     * @return {?}
     */
    getGlobalTargetProperty(target, prop) {
        return Promise.resolve();
    }
}
AdvancedRendererNoop.decorators = [
    { type: Injectable },
];
/**
 * @nocollapse
 */
AdvancedRendererNoop.ctorParameters = () => [];
function AdvancedRendererNoop_tsickle_Closure_declarations() {
    /** @type {?} */
    AdvancedRendererNoop.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    AdvancedRendererNoop.ctorParameters;
}
